### PR TITLE
ENH: Composite Reflectivity Function 

### PR DESCRIPTION
**Pull request by dopplerchase**
Hello pyart'ers, 

Since moving out west, the low level radar scans show lack luster echoes. Often times I use composite reflectivity on radar scope for quick looks at storms. One day, I wanted to plot composite Z with pyart, thinking there was a function to do this, but upon looking I could not find one. Hence I sought out to do this with some code. The trick I took here was sorting each sweeps azimuth such that the order of the matrix is the same for every single sweep. What this enables is the .max() call on the reordered matrix at the end.

Spot tests seem to work, but a full quantitative analysis has not been done.

### CODE REVIEW

File: tests/retrieve/test_comp_z.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+""" Unit Tests for Py-ART's retrieve/qvp.py module. """
+
+import numpy as np
+from numpy.testing import assert_equal,assert_array_equal
+
```
Comment (mgrover1) [inline tests/retrieve/test_comp_z.py:5]: I think you need another import copy here

**Comment by mgrover1** [comment]
@dopplerchase I think you need to add it to the `pyart/retrieve/__init__,py` file to expose that function

File: pyart/retrieve/comp_z.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+Calculate the composite reflectivity 
+
+"""
+
+import copy
+import numpy as np
+from netCDF4 import num2date
+from pandas import to_datetime
+
+def composite_reflectivity(radar,rhv_filter=True,rhv_value=0.95,verbose=False):
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:11]: I would make this more generic to take in any gatefilter, and if there isn't one, you can create one with the parameters you mentioned (for rho_hv)

File: pyart/retrieve/comp_z.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    
+    """
+    Composite Reflectivity 
+    
+    Often a legacy product, composite reflectivity is: 
+    "A display or mapping of the maximum radar reflectivity factor at any altitude as a function of position on the ground." - AMS Glossary
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:17]: Can you add citation here?

File: pyart/retrieve/comp_z.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    to the physical spinning of the radar antenna. 
+    
+    Author: Randy J. Chase (@dopplerchase) 
+    
+    Note: this function could potenially be parallelized at the loop, but for now runs in ~5 seconds on my 16 GB Mac
+    
+    Parameters
+    ----------
+    radar : Radar
+        Radar object used.
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:33]: specify pyart.core.Radar

File: pyart/retrieve/comp_z.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    rhv_filter : bool
+        boolean flag to turn on correlation coef. filtering. 
+    rhv_value : float
+        value to consider 'meteorological'
+    verbose : bool
+        boolean flag to turn the printing of radar tilts on or off. 
+
+    Returns
+    -------
+    comp_dict : Dictonary
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:43]: Can you have it return the field dict? So we can add it back to the radar object?

File: pyart/retrieve/comp_z.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    
+    #loop over all measured sweeps 
+    for sweep in radar.sweep_number['data']:
+    
+        #get start and stop index numbers
+        s_idx = radar.sweep_start_ray_index['data'][sweep]
+        e_idx = radar.sweep_end_ray_index['data'][sweep]+1
+        
+        #grab radar data 
+        z = radar.get_field(sweep,'reflectivity')
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:57]: Make sure to add a `reflectivity_field` argument so the user can specify this in the function

File: pyart/retrieve/comp_z.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        
+        #filter out values if desired. Future we should use gate filters...
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:59]: Yup! Making this a gate filter would be ideal

File: pyart/retrieve/comp_z.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        lat = lat[az_ids]
+        
+        #if the first sweep, store re-ordered lons/lats
+        if sweep == 0:
+            lon_0 = copy.deepcopy(lon)
+            lon_0[-1,:] = lon_0[0,:]
+            lat_0 = copy.deepcopy(lat)
+            lat_0[-1,:] = lat_0[0,:]
+        
+        #if 360 scan, upsample to super res
```
Comment (mgrover1) [inline pyart/retrieve/comp_z.py:85]: Why 720? Let's make this a parameter too - mentioning that default it will be at "super resolution" or 0.5 degrees

**Review by mgrover1 — CHANGES_REQUESTED** [review]
Thanks @dopplerchase for putting this together!! This PR looks great - just a few comments and feedback. Let me know if you need any help it, including an example!

It would be nice to have an example in this PR too.

I think users are going to be **very excited** about this new functionality. Super helpful to the community, and thanks for adding the tests.

File: pyart/retrieve/comp_z.py (hunk 5 above)
Comment (dopplerchase) [reply pyart/retrieve/comp_z.py:43]: Not quite because this is only 1 sweep. So the size of the matrix doesnt match any other field

Comment (mgrover1) [reply pyart/retrieve/comp_z.py:43]: Ohh good point... what if it returned a separate radar object? With a single field?

**Comment by mgrover1** [comment]
@dopplerchase - I can help with some of these changes if you'd like (especially integrating the gatefilter in here). Is that something you would be interested in?

**Comment by dopplerchase** [comment]
Yeah I have 0 experience implementing a gatefilter. It would be best if you can help with that

**Comment by mgrover1** [comment]
> Yeah I have 0 experience implementing a gatefilter. It would be best if you can help with that

Happy to help :)

File: pyart/retrieve/comp_z.py (hunk 5 above)
Comment (mgrover1) [reply pyart/retrieve/comp_z.py:43]: Nevermind! Using a dictionary would make sense and is consistent!

Comment (zssherman) [reply pyart/retrieve/comp_z.py]: Discussed this as the shape and info of the radar object changes, will go for radar object return.

File: pyart/retrieve/comp_z.py (hunk 7 above)
Comment (zssherman) [reply pyart/retrieve/comp_z.py]: done

**Comment by zssherman** [comment]
@dopplerchase With going the radar object route, with composite reflectivity, what typically is used for other fields for this data. Example, what would fixed angle be set to? Time array? etc just trying to figure out what to set these fields to, and not entirely familiar what is used as the defaults for a composite field.

**Comment by mgrover1** [comment]
@zssherman just let me know when this is ready for a review!

**Comment by zssherman** [comment]
@mgrover1 I still need to do the radar object return, was waiting on dopplerchase for the response on attributes and which values to use for time, azimuth etc, if we are doing one sweep we can return a radar.extract_sweeps.

**Comment by dopplerchase** [comment]
Hey sorry guys, busy time for me right now. As for the attributes, could you elaborate more on what you need? or point me to an example?

**Comment by zssherman** [comment]
@dopplerchase No worries, my question more pertained, if we return a radar object, for a composite radar, should the fixed angle be 0? elevations etc, wasn't sure how these values get defined for a composite field, or if we take the mean of these fields? Because one way, i this is one sweep, is just use the radar.extract_sweeps 0 sweep and have a radar set up for that, however if we are setting the azimuth using 360 or 720 if it will differ from the nrays of one sweep. Can we define the resolution from instrument parameters so it matches the original radar, or is it better to allow the user to set that?

**Comment by mgrover1** [comment]
@zssherman how can I help here? I think this would be great to include in the upcoming release!

**Comment by zssherman** [comment]
@mgrover1 I think the only thing left is plugging everything into a radar class similar to what's done in the ios like the nexrad or cfradial readers. I'm trying to remember what Scott said on Thursday of values to use. Think fixed angle as 0.

**Comment by mgrover1** [comment]
> @mgrover1 I think the only thing left is plugging everything into a radar class similar to what's done in the ios like the nexrad or cfradial readers. I'm trying to remember what Scott said on Thursday of values to use. Think fixed angle as 0.

Yeah! Fixed angle of 0, treating it as a single sweep of a radar

**Comment by zssherman** [comment]
@mgrover1 Tests failed due to a newer matplotlib version, I fixed it, but we will have to update the radarmapdisplay and gridmapdisplay plot_colorbar to make ax as a parameter and not optional.

**Comment by mgrover1** [comment]
Okay - @zssherman @dopplerchase this should be ready to go. I added in the example in the last commit, which resulted in the following image:

test_image

**Review by zssherman — APPROVED** [review]

**Comment by zssherman** [comment]
@mgrover1 only thing left i see is the reference

File: pyart/retrieve/comp_z.py (hunk 3 above)
Comment (mgrover1) [reply pyart/retrieve/comp_z.py]: On second thought, I don't see a well-defined (or citable) definition out there...

**Comment by mgrover1** [comment]
@zssherman after searching for a bit, I could not find a good reference. @dopplerchase any thoughts here? If we can't find one, I think we can go ahead and merge.

**Review by mgrover1 — APPROVED** [review]

**Pull request merged**